sharedbuffers/mapped_struct.py

```diff
@@ -1708,6 +1712,65 @@ def _c_search_hkey_ui64(hkey, pindex, stride0, length, hint):
 
     @cython.cfunc
     @cython.locals(
+        hkey = cython.longlong,
+        mkey = cython.longlong,
+        lo = cython.size_t, hi = cython.size_t, length = cython.size_t,
+        mid = cython.size_t, mid2 = cython.size_t, stride0 = cython.size_t, hint = cython.size_t,
+        pindex = cython.p_char, skip = cython.size_t)
+    @cython.returns(cython.size_t)
+    def _c_search_hkey_i64(hkey, pindex, stride0, length, hint):
+        hi = length
+        lo = 0
+        if lo < hi:
+            # First iteration a quick guess assuming uniform distribution of keys
+            mid = min(hint, hi-1)
+            mkey = cython.cast(cython.p_longlong, pindex + stride0 * mid)[0]
+            if mkey < hkey:
+                # Got a lo guess, now skip-search forward for a hi
+                lo = mid = mid+1
+                skip = 32
+                while skip > 0 and mid + skip < hi:
+                    if cython.cast(cython.p_longlong, pindex + stride0 * (mid+skip))[0] < hkey:
+                        lo = mid+1
+                        mid += skip
+                        skip *= 2
+                    else:
+                        hi = mid + skip
+                        break
+            elif mkey > hkey:
+                # Got a hi guess, now skip-search backwards for a lo
+                hi = mid
+                skip = 32
+                while skip > 0 and mid > lo + skip:
+                    if cython.cast(cython.p_longlong, pindex + stride0 * (mid-skip))[0] > hkey:
+                        hi = mid
+                        mid -= skip
+                        skip *= 2
+                    else:
+                        lo = mid - skip
+                        break
+            else:
+                # hit, but must find the first
+                # good idea to go sequential because we assume collisions are unlikely
+                while mid > lo and cython.cast(cython.p_longlong, pindex + stride0 * (mid-1))[0] == hkey:
+                    mid -= 1
+                return mid
+        # Final stretch: search the remaining range with a regular binary search
+        while lo < hi:
```

Comment:
If you add a hkey!=mkey here wouldn't you be able to do the while afterwards and prevent code duplication?
like:
```
if lo< hi:
    #first iteration
while mkey != hkey && lo<hi
    #bsearch
if mkey == hkey
    while mid > lo and cython.cast....:
         mid -=1
    return mid
return lo
```